### PR TITLE
feat(kuma-cp) Validate CA on Universal

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

Validate that CA can be set to provided or can be changed.

### CODE REVIEW

File: pkg/core/validators/apis/mesh/mesh_validator.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		}
+	}
+	return nil
+}
+
+func (m *MeshValidator) validateProvidedCaRoot(ctx context.Context, mesh string) error {
+	certs, err := m.ProvidedCaManager.GetSigningCerts(ctx, mesh)
+	if err != nil {
+		verr := validators.ValidationError{}
+		verr.AddViolation("mtls.ca.provided", "There are no provided CA for a given mesh")
```
Comment (yskopets) [inline pkg/core/validators/apis/mesh/mesh_validator.go]: This error message is not clear from a user perspective.

E.g., a user might think `I'm trying to change CA type to "provided" right now. So, I know that there was no provided CA before.`

File: pkg/core/validators/apis/mesh/mesh_validator.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		return verr.OrNil()
+	}
+	if len(certs) == 0 {
+		verr := validators.ValidationError{}
+		verr.AddViolation("mtls.ca.provided", "There are no signing certificate in provided CA for a given mesh")
```
Comment (yskopets) [inline pkg/core/validators/apis/mesh/mesh_validator.go]: This message will appear in the output of `kumactl`. Can we make it more actionable ?

File: pkg/core/validators/apis/mesh/mesh_validator.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package mesh
```
Comment (yskopets) [inline pkg/core/validators/apis/mesh/mesh_validator.go]: Let's move this file next to `mesh_manager.go`

File: pkg/core/managers/apis/mesh/mesh_manager.go [hunk 4]
```diff
@@ -146,6 +161,14 @@ func (m *meshManager) Update(ctx context.Context, resource core_model.Resource,
...
 	}
+
+	currentMesh := &core_mesh.MeshResource{}
+	if err := m.Get(ctx, currentMesh, core_store.GetBy(core_model.MetaToResourceKey(mesh.GetMeta()))); err != nil {
+		return err
+	}
+	if err := m.meshValidator.ValidateUpdate(ctx, currentMesh, mesh); err != nil {
+		return err
+	}
 	return m.store.Update(ctx, mesh, fs...)
```
Comment (yskopets) [inline pkg/core/managers/apis/mesh/mesh_manager.go]: Doesn't take into account optimistic lock (`Mesh` might change in between `m.meshValidator.ValidateUpdate` and `m.store.Update`)

File: pkg/core/managers/apis/mesh/mesh_manager_test.go [hunk 5]
```diff
@@ -83,9 +270,14 @@ var _ = Describe("Mesh Manager", func() {
 			Expect(store.IsResourceNotFound(err)).To(BeTrue())
 
 			// and built-in mesh CA is deleted
-			_, err = caManager.GetRootCerts(context.Background(), meshName)
+			_, err = builtinCaManager.GetRootCerts(context.Background(), meshName)
+			Expect(err).ToNot(BeNil())
+			Expect(err).To(MatchError("failed to load CA key pair for Mesh \"mesh-1\": Resource not found: type=\"Secret\" name=\"builtinca.mesh-1\" mesh=\"mesh-1\""))
```
Comment (yskopets) [inline pkg/core/managers/apis/mesh/mesh_manager_test.go]: Let's add a TODO to make behaviour of `builtin` and `provided` CAs consistent

File: pkg/core/validators/apis/mesh/mesh_validator.go (hunk 1 above)
Comment (jakubdyszkiewicz) [reply pkg/core/validators/apis/mesh/mesh_validator.go]: Changed, actually merged two of them to one message.

File: pkg/core/validators/apis/mesh/mesh_validator.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/core/validators/apis/mesh/mesh_validator.go]: Added info about kumactl cmd to run

File: pkg/core/managers/apis/mesh/mesh_manager.go (hunk 4 above)
Comment (yskopets) [reply pkg/core/managers/apis/mesh/mesh_manager.go]: Will be resolved in https://github.com/Kong/kuma/issues/487

**Pull request merged**